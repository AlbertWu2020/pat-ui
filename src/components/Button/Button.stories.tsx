import React from "react";
import { action } from "@storybook/addon-actions";
import Button, { ButtonType, ButtonSize } from "./Button";

export default {
  title: "Button",
  component: Button,
};

const buttonStyle: React.CSSProperties = {
  marginRight: "5px",
  marginTop: "5px",
};

export const DefaultButton = () => (
  <Button onClick={action("Default Button clicked")}>Default Button</Button>
);

export const DiffSizeButton = () => (
  <div>
    <Button
      style={buttonStyle}
      btnSize={ButtonSize.Small}
      onClick={action("Small Button clicked")}
    >
      Small Button
    </Button>
    <Button style={buttonStyle} onClick={action("Default Button clicked")}>
      Default Button
    </Button>
    <Button btnSize={ButtonSize.Large} onClick={action("Large Button clicked")}>
      Large Button
    </Button>
  </div>
);

export const DiffTypeButton = () => (
  <div>
    <Button
      style={buttonStyle}
      btnType={ButtonType.Primary}
      onClick={action("Primary Button clicked")}
    >
      Primary Button
    </Button>
    <Button
      style={buttonStyle}
      btnType={ButtonType.Secondary}
      onClick={action("Secondary Button clicked")}
    >
      Secondary Button
    </Button>
    <Button
      style={buttonStyle}
      btnType={ButtonType.Danger}
      onClick={action("Danger Button clicked")}
    >
      Danger Button
    </Button>
    <Button
      style={buttonStyle}
      btnType={ButtonType.Default}
      onClick={action("Default Button clicked")}
    >
      Default Button
    </Button>
    <Button
      style={buttonStyle}
      disabled
      btnType={ButtonType.Default}
      onClick={action("DisabledDefault Button clicked should not work")}
    >
      Disabled Default Button
    </Button>
    <Button
      style={buttonStyle}
      btnType={ButtonType.Link}
      onClick={action("Link Button clicked")}
    >
      Link Button
    </Button>
    <Button
      disabled
      btnType={ButtonType.Link}
      onClick={action("Disabled Link Button clicked should not work")}
    >
      Disabled Link Button
    </Button>
  </div>
);
