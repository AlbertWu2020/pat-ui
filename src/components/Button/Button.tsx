import React from 'react';
import {classNames} from '../../utils/classNames';

export enum ButtonSize {
  Large = 'lg',
  Small = 'sm',
}
export enum ButtonType {
  Primary = 'primary',
  Secondary = 'secondary',
  Danger = 'danger',
  Default = 'default',
  Link = 'link',
}
export interface IButtonProps {
  className?: string;
  btnSize?: ButtonSize;
  btnType?: ButtonType;
  children?: React.ReactNode;
  disabled?: boolean;
}

type NativeButtonProps = IButtonProps &
  React.ButtonHTMLAttributes<HTMLButtonElement>;
type NativeAchorButtonProps = IButtonProps &
  React.AnchorHTMLAttributes<HTMLAnchorElement>;
type PatButtonProps = NativeButtonProps | NativeAchorButtonProps;

const Button: React.FC<PatButtonProps> = (props) => {
  const {btnSize, btnType, children, disabled, className, ...rest} = props;
  let styleClasses = classNames('btn', {
    [`btn-${btnType}`]: true,
    [`btn-${btnSize}`]: !!btnSize,
    disabled: !!(disabled && btnType === ButtonType.Link),
  });
  if (className) {
    styleClasses += ' ' + className;
  }

  let btn;
  if (btnType !== ButtonType.Link) {
    btn = (
      <button
        className={styleClasses}
        disabled={disabled}
        {...(rest as NativeButtonProps)}
      >
        {props.children}
      </button>
    );
  } else {
    if (disabled) {
      rest.onClick = (e: React.MouseEvent) => {
        e.preventDefault();
      };
    }
    btn = (
      <a className={styleClasses} {...(rest as NativeAchorButtonProps)}>
        {props.children}
      </a>
    );
  }
  return btn;
};

Button.defaultProps = {
  btnType: ButtonType.Default,
  disabled: false,
};

export default Button;
