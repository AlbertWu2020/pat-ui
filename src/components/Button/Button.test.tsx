import React from 'react';

import {render, fireEvent} from '@testing-library/react';
import Button from './Button';

describe('Buttton', () => {
  it('should render default button', () => {
    const btnProps = {
      onClick: jest.fn(),
    };
    const wrapper = render(<Button {...btnProps}>Default Button</Button>);
    const element = wrapper.queryByText('Default Button') as HTMLElement;
    expect(element).toBeInTheDocument();
    expect(element.tagName).toBe('BUTTON');
    expect(element).toHaveClass('btn btn-default');
    expect(btnProps.onClick).toHaveBeenCalledTimes(0);
    fireEvent.click(element);
    expect(btnProps.onClick).toHaveBeenCalledTimes(1);
  });

  it('should render correct button based on different props', () => {});
});
