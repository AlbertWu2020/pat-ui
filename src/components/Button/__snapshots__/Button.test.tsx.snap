// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn btn-default"
  >
     Snapshot Button 
  </button>
</DocumentFragment>
`;

exports[`Buttton should match snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn btn-default"
  >
     Snapshot Button 
  </button>
</DocumentFragment>
`;
